### DISCUSSION TITLE
Lag Jacobian

### DISCUSSION THREAD

**Discussion by colinjcotter** [firedrakeproject/firedrake#2974] [Firedrake support]
Hi folks, 
  I'm solving a linear time dependent PDE with a coefficient, and want to do lots of long runs with different values of the coefficient. To make this fast, I'm doing snes_lag_jacobian -2, so it reuses the LU factorisation. However, this means changing the coefficient doesn't do anything. Is there a way I can force it to update from the new coefficient between the runs?

all the best
--cjc

**Comment by wence-** [comment]
> Hi folks, I'm solving a linear time dependent PDE with a coefficient, and want to do lots of long runs with different values of the coefficient. To make this fast, I'm doing snes_lag_jacobian -2, so it reuses the LU factorisation. However, this means changing the coefficient doesn't do anything. Is there a way I can force it to update from the new coefficient between the runs?

So you want it to be computed once in a solve (and reused for the entirety of that solve), but then the next solve you change the coefficient and so you want it recomputed? If so, then I think you want [`-snes_lag_jacobian_persists false`](https://petsc.org/main/manualpages/SNES/SNESSetLagJacobianPersists/)

**Comment by colinjcotter** [comment]
No, I want it to be computed for a set of timesteps, and then reset it for a next set of timesteps.

________________________________________

**Reply by wence-** [reply]
Ah, then just call `solver.snes.setLagJacobian(-2)` whenever you want it recomputed the next time snes calls `snescomputejacobian`.

**Reply by jaemoshin** [reply]
I've checked the attributes of snes and setLagJacobian does not seem to be an attribute.

**Reply by wence-** [reply]
I guess it is not yet exposed in petsc4py, you can probably figure it out by looking in `petsc/src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx`
